[HttpClient] Add support for amphp/http-client v5

nicolas-grekas opened the pull request.

nicolas-grekas committed:
[HttpClient] Add support for amphp/http-client v5

nicolas-grekas commented on composer.json:
```diff
@@ -121,9 +121,7 @@
     "require-dev": {
-        "amphp/amp": "^2.5",
-        "amphp/http-client": "^4.2.1",
-        "amphp/http-tunnel": "^1.0",
```
~removing for now to unlock installing amp v3~
looking forward to v2 :)

nicolas-grekas commented on src/Symfony/Component/HttpClient/AmpHttpClient.php:
```diff
@@ -158,8 +154,8 @@ public function reset(): void
-            foreach ($pushedResponses as [$pushedUrl, $pushDeferred]) {
-                $pushDeferred->fail(new CancelledException());
+            foreach ($pushedResponses as [$pushedUrl, $pushedResponse]) {
+                $pushedResponse->error(new CancelledException());
```
I'm unsure about this: how can I cancel a pushed response from outside the push-handler?

nicolas-grekas commented on src/Symfony/Component/HttpClient/Internal/AmpClientState.php:
```diff
@@ -203,13 +189,11 @@ private function handlePush(Request $request, Promise $response, array $options)
             'local_pk' => $options['local_pk'],
         ]];
-
-        return $deferred->promise();
```
this maps to my note on AmpHttpClient: how can I cancel a pushed response from outside the push-handler?

kelunik replied:
I need to take a look myself, but the equivalent to the previous version would be replacing `Deferred` with `DeferredFuture` and then this line with `$deferred->getFuture()->await()` and then call `error` on the deferred for rejection.

trowski replied:
Similar to before. Return the future from a `DeferredFuture` in the push handler, then later either resolve it with null to accept the push or fail it with `CancelledException` to reject the push.

@kelunik What do you think of passing a `DeferredCancellation` to the push handler to provide a more direct way to cancel the push? We already have an instance, so the question is do we want to expose it or not?

nicolas-grekas replied:
Thanks, I think I now get it. PR updated, see [3rd commit](https://github.com/nicolas-grekas/symfony/pull/43/commits/ebfd618abee97fc25f3d46ae70801c17b62e43d1)
Please let me know if I made a mistake.

nicolas-grekas reviewed (commented):
Here is a WIP aimed at adding support for AMPv3, aka fibers.
I didn't grasp my head around a few things, see review comments.
But the most critical thing I don't get for now is how to change the code in `AmpResponse::select()` and `AmpResponse::stopLoop()`.

/cc @kelunik would you be available to have a look and give me some hints maybe? :pray:

kelunik commented on src/Symfony/Component/HttpClient/Internal/AmpBody.php:
```diff
@@ -43,105 +44,101 @@ public function __construct($body, &$info, \Closure $onProgress)
     }
 
-    public function read(): Promise
+    public function read(Cancellation $cancellation = null): ?string
```
```suggestion
    public function read(?Cancellation $cancellation = null): ?string
```

nicolas-grekas replied:
that's Symfony's CS :)

kelunik commented on src/Symfony/Component/HttpClient/Response/AmpResponse.php:
```diff
@@ -315,7 +297,7 @@ private static function followRedirects(Request $originRequest, AmpClientState $
             try {
                 // Discard body of redirects
-                while (null !== yield $response->getBody()->read()) {
+                while (null !== $response->getBody()->read()) {
                 }
```
This can probably be `$response->getBody()->close()` now?

nicolas-grekas replied:
updated thanks

kelunik commented on src/Symfony/Component/HttpClient/Response/AmpResponse.php:
```diff
@@ -216,7 +195,7 @@ private static function select(ClientState $multi, float $timeout): int
         return null === self::$delay ? 1 : 0;
```
There shouldn't be any manual calls to `Loop::run()` and `Loop::stop()` anymore.

nicolas-grekas replied:
yes, but I'm struggling a bit with figuring out the correct replacement :)
something with suspension?

kelunik replied:
Likely suspensions, yes. Suspensions shouldn't be cached in a static property here, as they're fiber-specific. So if this is called from two different fibers, it'll break. Use a local variable instead.

kelunik commented on src/Symfony/Component/HttpClient/Response/AmpResponse.php:
```diff
@@ -346,7 +323,7 @@ private static function followRedirects(Request $originRequest, AmpClientState $
 
-            foreach ($originRequest->getRawHeaders() as [$name, $value]) {
+            foreach ($originRequest->getHeaderPairs() as [$name, $value]) {
                 $request->setHeader($name, $value);
```
This should likely call `addHeader` instead?

nicolas-grekas replied:
indeed, thanks

kelunik commented:
@nicolas-grekas Could you please do separate commits instead of force pushing for now? Makes it much easier to see what changed.

nicolas-grekas commented:
Sure, thanks for reviewing! I pushed something with a static suspension before seeing your comment and it breaks of course :)
I'm looking forward to your suggestions if you have more :)

nicolas-grekas commented:
I pushed something that might be closer to the working solution, but I'm not there yet.
Hints still welcomed @kelunik :pray:

nicolas-grekas commented on src/Symfony/Component/HttpClient/Response/AmpResponse.php:
```diff
@@ -152,7 +133,7 @@ public function __wakeup()
     {
         try {
-            $this->doDestruct();
+            //$this->doDestruct();
```
I figured out the blocker: we cannot switch fibers in destructors.
Most tests pass now. The ones that rely on destructors fail of course.

I tried this instead:
```php
            if ($this->initializer && null === $this->info['error']) {
                $clone = clone $this;
                $clone->shouldBuffer = true;
                EventLoop::queue($clone->getHeaders(...));
            }
```

It doesn't work (yet?) but I fear it cannot work because it relies on the event loop having a chance to be resumed after the destructor is called, which might not be always the case.

I might miss something of course.

nicolas-grekas replied:
I can make many tests pass by patching php and Amp this way.

I'm going to need your help @kelunik and @trowski to figure out if there is something to improve in either PHP, Amp and/or Symfony.


1. allow switching fibers in destructors:

```diff
--- a/Zend/zend_fibers.c
+++ b/Zend/zend_fibers.c
@@ -391,7 +391,7 @@ ZEND_API void zend_fiber_switch_unblock(void)
 
 ZEND_API bool zend_fiber_switch_blocked(void)
 {
-       return zend_fiber_switch_blocking;
+       return false;//zend_fiber_switch_blocking;
 }
```

2. ignore `Cannot suspend in a force-closed fiber`:

```diff
--- a/src/EventLoop/Internal/AbstractDriver.php
+++ b/src/EventLoop/Internal/AbstractDriver.php
@@ -538,6 +538,7 @@ abstract class AbstractDriver implements Driver
     private function createCallbackFiber(): void
     {
         $this->callbackFiber = new \Fiber(function (): void {
+            try {
             do {
                 $this->invokeMicrotasks();
 
@@ -599,6 +600,8 @@ abstract class AbstractDriver implements Driver
                 /** @noinspection PhpUnhandledExceptionInspection */
                 \Fiber::suspend($this->internalSuspensionMarker);
             } while (true);
+            } catch (\FiberError) {
+            }
         });
     }
```

trowski replied:
Allowing fiber switches in destructors requires the garbage collector to be safe for re-entry from multiple fibers.

It's also generally unadvisable to switch fibers in a finally block, which is likely what is causing the force-closed fiber issue. This is similar to force-closing a generator, you cannot yield in the finally block.

Perhaps the solution then is to move any I/O out of a finally block?

You can switch fibers in a shutdown function, in case that is of any help.

I took a short look at the code this morning and… wow there's a lot going on here through multiple layers. Wrapping `$this->doDestruct()` into `EventLoop::queue($this->doDestruct(...));` had some promising results, but throwing the exceptions from the destructor to the event loop handler. Perhaps wrap that call with a catch that adds the exceptions to a list and re-throws them at a later time? Can `initialize()` be run immediately upon receiving a response rather than waiting until the destructor to run if it wasn't otherwise run?

trowski replied:
Calling `initialize()` soon after creating an `AmpResponse` instance seems more promising, but I'm getting exceptions thrown in the wrong places in tests I think. I'm going to bow out here for now, I have a rather busy weekend ahead of me. Hopefully @nicolas-grekas you can run with these ideas and make some progress.

nicolas-grekas replied:
Thanks for having a look! Calling initialize means waiting for the headers of the response to come back. This means we cannot call it immediately since this would make everything synchronous.

A common pattern with the concurrency model of HttpClient is to stack requests in an array and let them complete when the array is released:

```php
$pool = [];
foreach ($urls as $url) {
	$pool[] = $client->request('GET', $url);
}
unset($pool); // either explicit or more commonly implicit on exit of a method
```

The only way to make this API work is by doing I/O in destructors I'm afraid.

> Wrapping $this->doDestruct() into EventLoop::queue($this->doDestruct(...)); had some promising results, but throwing the exceptions from the destructor to the event loop handler

That's the closest we can have now indeed. But this requires running inside the event loop, which is not the most common for HttpClient. Alternatively, this could work with a new method on the client to wait for requests in such a pool, but this would break the current API, and the current API has one major desired feature: if ppl forget about checking their responses, they will get an exception if a 4xx/5xx is returned. This also requires doing I/O in destructors, and cannot be solved by any extra method of course.

When not running in an event loop, I thought about pooling destructed responses in a queue to check on every other method call on the client/responses. This would have the quite unpleasant effect or running those destructed requests out of context (think call-stack). This also wouldn't work for the very last set of responses. Yes, shutdown function (or $client->reset() as called by eg Symfony DI) could provide this missing tick, but the DX looks terrible...

An alternative idea could be to run in some sort of fiber-less mode inside destructors - that'd be a sync mode I suppose. Is that something that AMP could be of any help for?

> Allowing fiber switches in destructors requires the garbage collector to be safe for re-entry from multiple fibers

Is it correct or not that calling destructors during collect cycles is the step that requires this reentry, and that calling them on refcount=0 doesn't? If correct, would it be possible to allow switching fibers in refcount=0 situations? This would partially fix the problem for HttpClient since response objects are rarely caught in ref cycles.

Of course, the next question is: would it be possible to make the GC safe for re-entry from multiple fibers? The current situation makes destructors kinda reintroduce the “What color is your function?” problem, which fibers were designed to avoid.

kelunik replied:
> ```php
> $pool = [];
> foreach ($urls as $url) {
> 	$pool[] = $client->request('GET', $url);
> }
> unset($pool); // either explicit or more commonly implicit on exit of a method
> ```

That's quite a strange design choice and pattern IMO, which is indeed not possible to implement right now with fibers. In general, destructors shouldn't really do any work, similar to constructors.

> The current situation makes destructors kinda reintroduce the “What color is your function?” problem, which fibers were designed to avoid.

Indeed, but it's avoidable if destructors don't do any work, which worked quite well for us so far.

nicolas-grekas replied:
I had a quick look at amphp/http-client v5 and you have a few destructors that do just that: i/o to e.g. close sockets (and deferring to `EventLoop::queue()` to do so). The fact that Symfony's client does more than just closing sockets is not a reason to dismiss its style of API. This should be fully supported by the engine.

I get that this might be significant work of course.

Do you want me to open an issue on php-src? Who could help on the implementation? @trowski would you? Who else?

trowski replied:
Please open an issue in php-src. I'd like to see support for suspensions in destructors as well. It's been on my (long) todo list, but now might be a good time to move it toward the top. Unfortunately I know little about the GC, so will probably need some help.

We've avoided suspending in destructors by using a defer in the event loop, but this probably only really works in a long running app (and runs the risk of never running). Closing a socket doesn't require suspending because there's no result to await, but it is I/O, so that's a fair criticism.

I glanced through the GC code today, and it looks to already guard against re-entry. The concern of course would be that if a destructor which suspended never again resumed it would block the GC from ever being run again. I'm not sure if this is acceptable or not.

MatTheCat committed:
[Validator] Simplify `NoSuspiciousCharactersValidator`

keradus committed:
chore: simplify PHP CS Fixer config

requires ^3.50 of Fixer

PhilETaylor committed:
[AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex

llupa committed:
[Security][Tests] Update functional tests to better reflect end-user scenarios

nicolas-grekas committed:
minor #54062 [Validator] Simplify `NoSuspiciousCharactersValidator` (MatTheCat)

This PR was merged into the 7.0 branch.

Discussion
----------

[Validator] Simplify `NoSuspiciousCharactersValidator`

| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | N/A
| License       | MIT

https://github.com/php/php-src/issues/10647 has been fixed in PHP 8.1.17. Now that Symfony requires PHP ≥ 8.2, we can avoid calling `Spoofchecker::isSuspicious` for every check by leveraging its `$errorCode` parameter.

Commits
-------

c24e3e7d5e [Validator] Simplify `NoSuspiciousCharactersValidator`

palgalik committed:
[Mailer][Brevo] Remove tags from mandatory event arguments

'tags' in not mandatory part of an event
therefore it got removed from the validation process

fabpot committed:
bug #54089 [Mailer] [Brevo] Remove tags from mandatory event arguments (palgalik)

This PR was merged into the 6.4 branch.

Discussion
----------

[Mailer] [Brevo] Remove tags from mandatory event arguments

'tags' in not mandatory part of an event
therefore it got removed from the validation process

| Q             | A
| ------------- | ---
| Branch?       |  6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix [54088](https://github.com/symfony/symfony/issues/54088)
| License       | MIT

Commits
-------

dfb41cda2d [Mailer][Brevo] Remove tags from mandatory event arguments

trowski commented:
Once https://github.com/php/php-src/pull/13460 is merged (presumably for PHP 8.4) I expect this to work as intended. Looking forward to being able to fiber-switch in destructors in AMPHP as well. 😄

nicolas-grekas committed:
[HttpClient] Fix deprecation on PHP 8.3

nicolas-grekas commented:
@trowski I was just playing with that PR and it looks promising! (I found https://github.com/amphp/socket/issues/110 on your side in the process ;) )

fabpot committed:
minor #54073 chore: simplify PHP CS Fixer config (keradus)

This PR was merged into the 7.1 branch.

Discussion
----------

chore: simplify PHP CS Fixer config

requires ^3.50 of Fixer

| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix complex config
| License       | MIT

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

Commits
-------

7e62355d32 chore: simplify PHP CS Fixer config

derrabus committed:
bug #54102 [HttpClient] Fix deprecation on PHP 8.3 (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[HttpClient] Fix deprecation on PHP 8.3

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Commits
-------

9407a1b07e [HttpClient] Fix deprecation on PHP 8.3

lyrixx committed:
[Mailer] Simplify MailerTestCommand

nicolas-grekas committed:
bug #54079 [AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex (PhilETaylor)

This PR was squashed before being merged into the 6.4 branch.

Discussion
----------

[AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #54078
| License       | MIT

Fixes regression in regex, fix provided by `@nicolas`-grekas in #54078

Commits
-------

ae16b2d49d [AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex

smnandre committed:
[AssetMapper] Throw exception in Javascript compiler when PCRE error

smnandre committed:
Remove undefined variable

The $path variable was originally removed in this commit, https://github.com/symfony/symfony/commit/c80a1abf8b448e4444448032d79764b9e8aeefdb#diff-ac1411bcebc4a2a31db403a027cb257f5a5c0e06936516435b88235149234a0fR73 However, it appears to have been reintroduced during subsequent merges

This glitch is not there in 6.4

smnandre committed:
[AssetMapper] Deprecate unused method `splitPackageNameAndFilePath`

ruudk committed:
[Clock] Add PHPUnit 10 attributes

PHPUnit 11 will complain with the following deprecation:
```
Metadata in doc-comments is deprecated and will no longer be supported in PHPUnit 12.
Update your test code to use attributes instead.
```

If we add the attribute too, it stops complaining. So this most be a way to support PHPUnit 11 and
older versions.

nicolas-grekas committed:
bug #54129 [Clock] Add attributes to support PHPUnit 10 + 11 (ruudk)

This PR was merged into the 6.4 branch.

Discussion
----------

[Clock] Add attributes to support PHPUnit 10 + 11

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

While upgrading to PHPUnit 11 I noticed the following deprecation:

> Metadata in doc-comments is deprecated and will no longer be supported in PHPUnit 12. Update your test code to use attributes instead.

If we add the attribute too, it stops complaining. So this most be a way to support PHPUnit 11 and older versions.

Commits
-------

0478d5467d [Clock] Add PHPUnit 10 attributes

chalasr committed:
minor #54086 [Security][Tests] Update functional tests to better reflect end-user scenarios (llupa)

This PR was squashed before being merged into the 5.4 branch.

Discussion
----------

[Security][Tests] Update functional tests to better reflect end-user scenarios

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | N/A
| License       | MIT

Pinging `@wouterj`

This PR is related to https://github.com/symfony/symfony/pull/53851 's Context.

> A person going through Symfony docs for the first time wanted to create their own LoginFormType as a next step in their learning Symfony journey and noticed that you can submit empty username/password with form login.
>
> They wanted to disallow this and tried to add validation. To validate a login form is not so straight forward as it either needs to be done with a custom authenticator (complex validation) or user provider if the data checks are simple.

Following comments:

https://github.com/symfony/symfony/pull/53851#issuecomment-1937763662
> Given the broken high-deps build, I wonder if this shouldn't even be done with a deprecation notice before making it throw in 8.0?

https://github.com/symfony/symfony/pull/53851#issuecomment-1938289588
> These are 3 tests submitting an empty login form to trigger a CSRF token error. This new condition now takes precedence, meaning it returns the wrong error.
I don't think that is something we have to worry about (in both situations, login errors), it rather reveals a bad test in our codebase. I can't think of a use-case that would result in success and will become a failure after this merge.

https://github.com/symfony/symfony/pull/53851#issuecomment-1938596547
> I think we need consensus on whether we find this a hard BC break that deserves a smooth upgrade path, but the test need to be fixed whatever the conclusion

Commits
-------

4155f66e7f [Security][Tests] Update functional tests to better reflect end-user scenarios

wouterj committed:
Merge branch '5.4' into 6.4

* 5.4:
  [Security][Tests] Update functional tests to better reflect end-user scenarios
  [HttpClient] Fix deprecation on PHP 8.3

wouterj committed:
Merge branch '6.4' into 7.0

* 6.4:
  [Security][Tests] Update functional tests to better reflect end-user scenarios
  [Clock] Add PHPUnit 10 attributes
  [AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex
  [HttpClient] Fix deprecation on PHP 8.3
  [Mailer][Brevo] Remove tags from mandatory event arguments

wouterj committed:
Merge branch '7.0' into 7.1

* 7.0:
  [Security][Tests] Update functional tests to better reflect end-user scenarios
  [Clock] Add PHPUnit 10 attributes
  [AssetMapper] Fix `JavaScriptImportPathCompiler` regression in regex
  [HttpClient] Fix deprecation on PHP 8.3
  [Mailer][Brevo] Remove tags from mandatory event arguments
  [Validator] Simplify `NoSuspiciousCharactersValidator`

fabpot committed:
feature #54125 [AssetMapper] Deprecate unused method `splitPackageNameAndFilePath` (smnandre)

This PR was squashed before being merged into the 7.1 branch.

Discussion
----------

[AssetMapper] Deprecate unused method `splitPackageNameAndFilePath`

| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Issues        | Fix #54123
| License       | MIT

This method is not used in the codebase, and duplicates` ImportMapEntry::splitPackageNameAndFilePath()`

Commits
-------

9af1f34417 [AssetMapper] Deprecate unused method `splitPackageNameAndFilePath`

Jibbarth committed:
[HttpClient] Preserve float in JsonMockResponse

fabpot committed:
minor #54122 [AssetMapper] Remove undefined $path variable (smnandre)

This PR was merged into the 7.0 branch.

Discussion
----------

[AssetMapper] Remove undefined $path variable

| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #...
| License       | MIT

The `$path` variable was originally removed in this commit, https://github.com/symfony/symfony/commit/c80a1abf8b448e4444448032d79764b9e8aeefdb#diff-ac1411bcebc4a2a31db403a027cb257f5a5c0e06936516435b88235149234a0fR73 However, it appears to have been reintroduced during subsequent merges.

The variable is not in the [6.4 code](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/AssetMapper/ImportMap/ImportMapConfigReader.php)

Commits
-------

5f959e2023 Remove undefined variable

fabpot committed:
bug #54113 [AssetMapper] Throw exception in Javascript compiler when PCRE error (smnandre)

This PR was squashed before being merged into the 6.4 branch.

Discussion
----------

[AssetMapper] Throw exception in Javascript compiler when PCRE error

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #...
| License       | MIT

`preg_match_callback` can return null when a PCRE error occured, leading there to a TypeError.

Let's throw an exception and expose the error message.

(follows #54078 / complementary to #54079 )

Commits
-------

2333b58370 [AssetMapper] Throw exception in Javascript compiler when PCRE error

fabpot committed:
Merge branch '6.4' into 7.0

* 6.4:
  [AssetMapper] Throw exception in Javascript compiler when PCRE error

fabpot committed:
Merge branch '7.0' into 7.1

* 7.0:
  [AssetMapper] Throw exception in Javascript compiler when PCRE error
  Remove undefined variable

fabpot committed:
minor #54110 [Mailer] Simplify MailerTestCommand (lyrixx)

This PR was submitted for the 6.4 branch but it was merged into the 7.1 branch instead.

Discussion
----------

[Mailer] Simplify MailerTestCommand

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

Should I target SF 7.1?

Commits
-------

fdb0340d36 [Mailer] Simplify MailerTestCommand

fabpot committed:
Update CHANGELOG for 5.4.37

fabpot committed:
Update VERSION for 5.4.37

fabpot committed:
Update CONTRIBUTORS for 5.4.37

fabpot committed:
Merge pull request #54157 from fabpot/release-5.4.37

released v5.4.37

fabpot committed:
Bump Symfony version to 5.4.38

fabpot committed:
Update CHANGELOG for 6.4.5

fabpot committed:
Update VERSION for 6.4.5

fabpot committed:
Merge pull request #54158 from fabpot/release-6.4.5

released v6.4.5

fabpot committed:
Bump Symfony version to 6.4.6

fabpot committed:
Update CHANGELOG for 7.0.5

fabpot committed:
Update VERSION for 7.0.5

fabpot committed:
Merge pull request #54159 from fabpot/release-7.0.5

released v7.0.5

fabpot committed:
Bump Symfony version to 7.0.6

derrabus committed:
Merge branch '5.4' into 6.4

* 5.4:
  Bump Symfony version to 5.4.38
  Update VERSION for 5.4.37
  Update CONTRIBUTORS for 5.4.37
  Update CHANGELOG for 5.4.37

derrabus committed:
bug #54146 [HttpClient] Preserve float in JsonMockResponse (Jibbarth)

This PR was merged into the 6.4 branch.

Discussion
----------

[HttpClient] Preserve float in JsonMockResponse

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Hello,

I am mocking an API Response that can send float in values :

```
{
    "values": 1.0
}
```

Converting this to a php array and giving it to JsonMockResponse return me an int after, because the missing json_encode flag to keep zero fraction, and leds to issue when I want to deserialize it into a float property.

In my mind it's a bugfix, but if it's considered as a new feature, I can target 7.1.

Commits
-------

4d120b8488 [HttpClient] Preserve float in JsonMockResponse

derrabus committed:
Merge branch '6.4' into 7.0

* 6.4:
  Bump Symfony version to 6.4.6
  Update VERSION for 6.4.5
  Update CHANGELOG for 6.4.5
  Bump Symfony version to 5.4.38
  Update VERSION for 5.4.37
  Update CONTRIBUTORS for 5.4.37
  Update CHANGELOG for 5.4.37
  [HttpClient] Preserve float in JsonMockResponse

derrabus committed:
Merge branch '7.0' into 7.1

* 7.0:
  Bump Symfony version to 7.0.6
  Update VERSION for 7.0.5
  Update CHANGELOG for 7.0.5
  Bump Symfony version to 6.4.6
  Update VERSION for 6.4.5
  Update CHANGELOG for 6.4.5
  Bump Symfony version to 5.4.38
  Update VERSION for 5.4.37
  Update CONTRIBUTORS for 5.4.37
  Update CHANGELOG for 5.4.37
  [HttpClient] Preserve float in JsonMockResponse

nicolas-grekas commented:
PR submitted at https://github.com/symfony/symfony/pull/54179
Thank you all for the review, let continue there.
